snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java

```diff
@@ -114,17 +165,31 @@ public void writeBandRasterData(Band sourceBand,
         checkSourceRegionInsideBandRegion(sourceWidth, sourceBandWidth, sourceHeight, sourceBandHeight, sourceOffsetX,
                                           sourceOffsetY);
         final ImageOutputStream outputStream = getOrCreateImageOutputStream(sourceBand);
-        long outputPos = (long) sourceOffsetY * (long) sourceBandWidth + sourceOffsetX;
-        pm.beginTask("Writing band '" + sourceBand.getName() + "'...", 1);//sourceHeight);
-        try {
-            final long max = sourceHeight * sourceWidth;
-            for (int sourcePos = 0; sourcePos < max; sourcePos += sourceWidth) {
-                sourceBuffer.writeTo(sourcePos, sourceWidth, outputStream, outputPos);
-                outputPos += sourceBandWidth;
+        final boolean fullRaster = isFullRaster(sourceWidth, sourceHeight, sourceBandWidth, sourceBandHeight);
+        if (useCache && (!fullRaster)) {
+            final VariableCache variableCache = writeCache.get(sourceBand);
+            final boolean canWrite = variableCache.update(sourceOffsetX, sourceOffsetY, sourceWidth, sourceHeight, sourceBuffer);
+            if (canWrite) {
+                synchronized (outputStream) {
+                    variableCache.writeCompletedBlocks(outputStream);
+                }
             }
             pm.worked(1);
```

Comment:
Not valid without pm.beginTask(...).
Also it should be wrapped with a try/finally block. Place pm.done() in the finally block.

Reply:
resolved - corrected progressbar handling